src/offline/Downloader.js

```diff
@@ -1,12 +1,186 @@
+    const layersMetadata = [];
+    /**
+     * @param {ol.layer.Base} layer .
+     * @param {Array<ol.layer.Group>} ancestors .
+     * @return {boolean} continue traversal
+     */
+    const visitLayer = (layer, ancestors) => {
+      console.log('Traversing layer', layer.getProperties());
+      const extent = userExtent; // could be different depending on type of layer (ex: background layers)
+      const minZoom = 0;
+      const maxZoom = 5; // FIXME
+      layersMetadata.push({
+        map,
+        minZoom,
+        maxZoom,
+        extent,
+        layer,
+        ancestors
+      });
+      return true;
+    };
+    map.getLayers().forEach((root) => {
+      utils.traverseLayer(root, [], visitLayer);
+    });
+    return layersMetadata;
+  }
+
+  queueLayer_(layerMetadata, queue) {
+    const layer = layerMetadata.layer;
+    if (layer instanceof ol.layer.Vector) {
+      this.queueVectorLayer_(layerMetadata, queue);
+    } else if (layer instanceof ol.layer.Tile || layer instanceof ol.layer.Image) {
+      this.queueLayerTiles_(layerMetadata, queue);
+    } else {
+      console.log('Unsupported layer', layerMetadata, layer.constructor.name);
+    }
+  }
+
+  queueVectorLayer_(layerMetadata, queue) {
+    console.log('Implement vector layer download');
+  }
+
+  sourceImageWMSToTileWMS_(source, projection) {
+    if (source instanceof ol.source.ImageWMS && source.getUrl() && source.getImageLoadFunction() === defaultImageLoadFunction) {
+      const tileGrid = ol.tilegrid.getForProjection(source.getProjection() || projection);
+      source = new ol.source.TileWMS({
+        url: source.getUrl(),
+        tileGrid: tileGrid,
+        attributions: source.getAttributions(),
+        projection: source.getProjection(),
+        params: source.getParams()
+      });
+    }
+    return source;
+  }
+
+
+  queueLayerTiles_(layerMetadata, queue) {
+    const {map, layer, extent, minZoom, maxZoom} = layerMetadata;
+
+    const projection = map.getView().getProjection();
+    const source = this.sourceImageWMSToTileWMS_(layer.getSource(), projection);
+    if (!source) {
+      return;
+    }
+    const tileGrid = source.getTileGrid();
+    const tileUrlFunction = source.getTileUrlFunction();
+
+    for (let z = minZoom; z < maxZoom; ++z) {
+      const queueByZ = [];
+      let minX, minY, maxX, maxY;
+      tileGrid.forEachTileCoord(extent, z, (tileCoord) => {
+        maxX = tileCoord[1];
+        maxY = tileCoord[2];
+        if (minX === undefined) {
+          minX = tileCoord[1];
+          minY = tileCoord[2];
+        }
+        const tile = {
+          coord: tileCoord,
+          url: tileUrlFunction(tileCoord, ol.has.DEVICE_PIXEL_RATIO, projection)
+        };
+        queueByZ.push(tile);
+      });
+
+      const centerTileCoord = [z, (minX + maxX) / 2, (minY + maxY) / 2];
+      queueByZ.sort((a, b) => magnitude2(a.coord, centerTileCoord) - magnitude2(b.coord, centerTileCoord));
+      queue.push(...queueByZ);
```

Comment:
... ? I don't know this notation, what's this ?

Reply:
This is the rest argument. It means, push every element of the array.

Reply:
thx

Reply:
called `spread` es6